_posts/2020-05-15-Listen-to-what-your-tests-are-mocking.md

```diff
@@ -0,0 +1,151 @@
+---
+title: A Complete Guide to Testing Your Software, Part 2 
+author: Alastair Smith
+category: testing
+created: 1589571968
+tags:
+ - testing
+ - craft
+ - TDD
+ - design
+ - ports and adapters
+---
+
+## <small><i>or, Listen to what your tests are mocking</i></small>
+
+You might remember from [last time](/testing/2020/04/26/complete-guide-testing-your-software-part-1.html) that we
+briefly covered the concept of [Hexagonal
+Architecture](https://en.wikipedia.org/wiki/Hexagonal_architecture_(software)), or "ports and adapters" as it's
+sometimes otherwise known, in the context of using mocks and stubs to resolve some of the pain of integrated tests. The
+"mocks mocking you" meme is a vein well-tapped at this point, so I wanted to come at it from a slightly different angle:
+listening to the tests we write, for what they're telling us about our design. Let's take a look at an example:
+
+```csharp
+public class Recorder
+{
+    private static readonly ILogger _logger;
+
+    public Recorder(ILogger logger) => _logger = logger;
+
+    public void Record(bool isA)
+    {
+        if (isA)
+        {
+            RecordA();
+            return;
+        }
+
+        RecordB();
+    }
+
+    private void RecordA()
+    {
+        _logger.LogInformation("A happened");
+    }
+
+    private void RecordB()
+    {
+        _logger.LogInformation("B happened");
+    }
+}
+```
+
+This is contrived code for sure, but the overall shape is common: consider a log statement in the success case or a
+different log statement in the error case; or registering a discount for a purchase because it's February and the
+customer was born in a leap year, or not. Paint your own domain over this structure, and then ask yourself the question
```

Comment:
`or registering a discount for a purchase because it's February and the
customer was born in a leap year, or not. Paint your own domain over this structure` 

This is sort of rambly.

Reply:
Resolved (hopefully) using bullet points to separate the examples.